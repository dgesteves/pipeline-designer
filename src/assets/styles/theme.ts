import "styled-components";
import { DefaultTheme } from "styled-components";

const theme: DefaultTheme = {
  colors: {
    white: "#ffffff",
    background: "#1D1E22",
    violet: "#c43ad6",
    turquoise: "#30CFD0",
  },
  sizes: {
    size_1: "0.625rem",
    size_2: "0.75rem",
    size_3: "0.875rem",
    size_4: "1rem",
    size_5: "1.125rem",
    size_6: "1.25rem",
    size_7: "1.5rem",
    size_8: "1.875rem",
    size_9: "2rem",
  },
  gradients: {
    textGradient: "linear-gradient(to right, #30CFD0, #c43ad6)",
  },
  shadows: {
    textShadow:
      "0 0 80px rgb(192 219 255 / 35%), 0 0 32px rgb(65 120 255 / 24%)",
  },
};

export default theme;
