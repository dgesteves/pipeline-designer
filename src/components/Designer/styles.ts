import styled from "styled-components";

export const DesignerSection = styled.div`
  display: flex;
  flex-direction: column;
  gap: ${({ theme }) => theme.sizes.size_4};
  position: fixed;
  color: ${({ theme }) => theme.colors.white};
  border: 2px double transparent;
  background-image: linear-gradient(
      ${({ theme }) => theme.colors.background},
      ${({ theme }) => theme.colors.background}
    ),
    radial-gradient(
      circle at left top,
      ${({ theme }) => theme.colors.turquoise},
      ${({ theme }) => theme.colors.violet}
    );
  background-origin: border-box;
  background-clip: padding-box, border-box;
  padding: ${({ theme }) => theme.sizes.size_4};
  border-radius: 8px;
`;

export const Option = styled(DesignerSection)`
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  position: static;
  color: ${({ theme }) => theme.colors.white};
  text-transform: uppercase;

  svg {
    cursor: grab;
  }
`;
