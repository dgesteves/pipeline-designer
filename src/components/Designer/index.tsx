import { DESIGNER_OPTIONS } from "../../constants";
import Draggable from "../../assets/icons/Draggable";
import { DesignerSection, Option } from "./styles";
import { DragEvent, useCallback } from "react";
import { TName } from "../../services/types";

export default function Designer() {
  const onDragStart = useCallback(
    (e: DragEvent<HTMLDivElement>, name: TName) => {
      e.dataTransfer.setData("name", name);
    },
    []
  );

  return (
    <DesignerSection data-testid="designer-test">
      {DESIGNER_OPTIONS.map(({ name, id }) => (
        <Option
          data-testid={`draggable-option`}
          key={id}
          draggable
          onDragStart={(e) => onDragStart(e, name)}
        >
          <p>{name}</p>
          <Draggable />
        </Option>
      ))}
    </DesignerSection>
  );
}
