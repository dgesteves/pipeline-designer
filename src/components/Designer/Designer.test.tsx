import { fireEvent, render, screen } from "@testing-library/react";
import Designer from "./index";
import { ThemeProvider } from "styled-components";
import theme from "../../assets/styles/theme";
import GlobalStyles from "../../assets/styles/globalStyles";

describe("Designer component", () => {
  it("should match previous snapshot", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Designer />
      </ThemeProvider>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render without crashing", () => {
    const { container } = render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Designer />
      </ThemeProvider>
    );

    expect(container).toBeInTheDocument();
  });

  it("should render the correct number of options", () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Designer />
      </ThemeProvider>
    );

    expect(screen.getAllByTestId("draggable-option").length).toBe(2);
    expect(screen.getAllByTestId("draggable-option").at(0)?.textContent).toBe(
      "retriever"
    );
    expect(screen.getAllByTestId("draggable-option").at(1)?.textContent).toBe(
      "reader"
    );
  });

  it("should call setData with correct data onDragStart", () => {
    render(
      <ThemeProvider theme={theme}>
        <GlobalStyles />
        <Designer />
      </ThemeProvider>
    );

    const draggableOption = screen.getAllByTestId("draggable-option")[0];
    const mockData = { setData: jest.fn() };

    fireEvent.dragStart(draggableOption, {
      dataTransfer: mockData,
    });

    expect(mockData.setData).toHaveBeenCalledWith("name", "retriever");
  });
});
