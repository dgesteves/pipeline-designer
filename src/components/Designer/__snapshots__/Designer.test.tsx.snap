// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Designer component should match previous snapshot 1`] = `
<div>
  <div
    class="sc-bczRLJ fOzJiW"
    data-testid="designer-test"
  >
    <div
      class="sc-bczRLJ sc-gsnTZi fOzJiW bIYcIQ"
      data-testid="draggable-option"
      draggable="true"
    >
      <p>
        retriever
      </p>
      <svg
        height="1em"
        viewBox="0 0 32 32"
        width="1em"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6h4v4h-4zM18 6h4v4h-4zM10 14h4v4h-4zM18 14h4v4h-4zM10 22h4v4h-4zM18 22h4v4h-4z"
          style="fill: currentColor;"
        />
        <path
          d="M0 0h32v32H0z"
          style="fill: none;"
        />
      </svg>
    </div>
    <div
      class="sc-bczRLJ sc-gsnTZi fOzJiW bIYcIQ"
      data-testid="draggable-option"
      draggable="true"
    >
      <p>
        reader
      </p>
      <svg
        height="1em"
        viewBox="0 0 32 32"
        width="1em"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M10 6h4v4h-4zM18 6h4v4h-4zM10 14h4v4h-4zM18 14h4v4h-4zM10 22h4v4h-4zM18 22h4v4h-4z"
          style="fill: currentColor;"
        />
        <path
          d="M0 0h32v32H0z"
          style="fill: none;"
        />
      </svg>
    </div>
  </div>
</div>
`;
